src/network/resolver.rs

```diff
@@ -0,0 +1,89 @@
+use std::{
+    collections::HashMap,
+    net::Ipv4Addr,
+    sync::{Arc, Mutex},
+    thread::{Builder, JoinHandle},
+};
+use tokio::{
+    runtime::Runtime,
+    sync::mpsc::{self, Sender},
+};
+use trust_dns_resolver::AsyncResolver;
+
+pub type Result<T> = std::result::Result<T, failure::Error>;
+pub type IpTable = Arc<Mutex<HashMap<Ipv4Addr, String>>>;
+
+const CHANNEL_SIZE: usize = 1_000;
+
+/// Lookup translates between IP addresses and hostnames asynchronously.
+pub trait Lookup: Send {
+    /// Lookup the hostname for the given IP address asynchronously.
+    fn lookup(&mut self, addr: Ipv4Addr);
+}
+
+/// An asynchronous implementation of the Lookup trait.
+pub struct Resolver {
+    tx: Option<Sender<Ipv4Addr>>,
+    handle: Option<JoinHandle<()>>,
+}
+
+impl Resolver {
+    /// Create a new resolver with the given address translation table.
+    pub fn new(table: IpTable) -> Result<Self> {
+        let runtime = Runtime::new()?;
+        let (tx, mut rx) = mpsc::channel::<Ipv4Addr>(CHANNEL_SIZE);
+        let (resolver, background) = AsyncResolver::from_system_conf()?;
```

Comment:
This is a little nitpicky of me: could we rename "resolver" here to something else (eg. dns_reverse_lookup or something similar)? I feel it will make the distinction between it and the struct in which it lives clearer. What do you think?

Reply:
Indeed, we could name it differently. I would suggest maybe `dns_lookup` without the `reverse` part, mainly because it can do more than reverse lookup, even if we only use that functionality here.